### PR TITLE
Refactor sqlthread

### PR DESCRIPTION
**Pull request by cis-muzahid**

### CODE REVIEW

File: stdeb.cfg [hunk 1]
```diff
@@ -1,7 +1,7 @@
 [DEFAULT]
 Package: pybitmessage
 Section: net
-Build-Depends: dh-python, libssl-dev, python-all-dev, python-setuptools
+Build-Depends: dh-python, libssl-dev, python-all-dev, python-setuptools, python-six
```
Comment (PeterSurda) [inline stdeb.cfg:4]: this should be a separate PR

File: src/helper_startup.py [hunk 2]
```diff
@@ -29,6 +25,8 @@
 except ImportError:
     get_plugin = None
 
+# pylint: disable=too-many-branches,too-many-statements
+
```
Comment (PeterSurda) [inline src/helper_startup.py:29]: is this a code quality only change? Then it should go into a separate PR.

**Review by PeterSurda — CHANGES_REQUESTED** [review]
Too big, can't review. Needs to be split. Separate code quality, support functions, and the individual levels.

File: stdeb.cfg (hunk 1 above)
Comment (g1itch) [reply stdeb.cfg:4]: It is 16a1177 merged from v0.6.

Comment (cis-muzahid) [reply stdeb.cfg]: Is it fine now?

**Comment by PeterSurda** [comment]
The upgrade design is too complicated. There is no need for a separate function per level.

File: src/class_sqlThread.py [hunk 3]
```diff
@@ -29,22 +18,120 @@
...
+        self.max_level = 11
+
+        # call upgrading level in loop
+        for l in range(self.current_level, self.max_level):
+            if int(l) == 3:
+                continue
+            self.upgrade_one_level(l)
+            self.run_migrations(l)
+
+    def upgrade_schema_data_level(self, level):
```
Comment (PeterSurda) [inline src/class_sqlThread.py:114]: ```def increment_settings_version``

File: src/class_sqlThread.py [hunk 4]
```diff
@@ -29,22 +18,120 @@
...
+
+        if level != self.get_current_level():
+            return None
+
+        # Migrate Db with level
+        method_name = 'upgrade_schema_data_' + str(level)
+        method = getattr(self, method_name, lambda: "Invalid version")
+        return method()
+
+    def run_migrations(self, file_name):
```
Comment (PeterSurda) [inline src/class_sqlThread.py:80]: ```def _upgrade_one_level_sql_statement```

File: src/class_sqlThread.py [hunk 5]
```diff
@@ -29,22 +18,120 @@
...
+    """
 
-class sqlThread(threading.Thread):
+    conn, cur = connection_build()
+
+    def __init__(self):
+        self.current_level = None
+        self.max_level = 11
+
+    def get_current_level(self):
```
Comment (PeterSurda) [inline src/class_sqlThread.py:58]: ```def __get_current_settings_version```
Maybe this can't be private, that's ok.

File: src/class_sqlThread.py [hunk 6]
```diff
@@ -29,22 +18,120 @@
...
+        """
+            Upgrade Db with respect to their versions
+        """
+        item = '''SELECT value FROM settings WHERE key='version';'''
+        parameters = ()
+        self.cur.execute(item, parameters)
+        return int(self.cur.fetchall()[0][0])
+
+    def upgrade_one_level(self, level):
```
Comment (PeterSurda) [inline src/class_sqlThread.py:67]: ```def _upgrade_one_level_method```

**Review by PeterSurda — CHANGES_REQUESTED** [review]

File: src/class_sqlThread.py [hunk 7]
```diff
@@ -102,11 +189,20 @@ def run(self):  # pylint: disable=too-many-locals, too-many-branches, too-many-s
         settingsversion = BMConfigParser().getint(
             'bitmessagesettings', 'settingsversion')
 
+        # Process SQL queries first20bytesofencryptedmessage in table inventory
```
Comment (PeterSurda) [inline src/class_sqlThread.py:192]: this section can probably be deleted

File: src/class_sqlThread.py [hunk 8]
```diff
@@ -102,11 +189,20 @@ def run(self):  # pylint: disable=too-many-locals, too-many-branches, too-many-s
...
+        if settingsversion == 1:
+            item = '''ALTER TABLE inventory ADD first20bytesofencryptedmessage blob DEFAULT '' '''
+            parameters = ()
+            self.cur.execute(item, parameters)
+            item = '''update settings set value=? WHERE key='version';'''
+            parameters = (2,)
+            self.cur.execute(item, parameters)
+
         # People running earlier versions of PyBitmessage do not have the
```
Comment (PeterSurda) [inline src/class_sqlThread.py:201]: lines 201-227 can be made into a separate method

Comment (PeterSurda) [reply src/class_sqlThread.py:201]: lines 227-292 can be split into 2 or 3 methods.

Comment (cis-muzahid) [reply src/class_sqlThread.py]: Done, please check the method name.

File: src/class_sqlThread.py (hunk 7 above)
Comment (cis-muzahid) [reply src/class_sqlThread.py]: Done

File: src/class_sqlThread.py (hunk 6 above)
Comment (cis-muzahid) [reply src/class_sqlThread.py]: Done

File: src/class_sqlThread.py (hunk 5 above)
Comment (cis-muzahid) [reply src/class_sqlThread.py]: Done

File: src/class_sqlThread.py (hunk 4 above)
Comment (cis-muzahid) [reply src/class_sqlThread.py]: Done

File: src/class_sqlThread.py (hunk 3 above)
Comment (cis-muzahid) [reply src/class_sqlThread.py]: Done

File: src/tests/test_sqlthread.py [hunk 9]
```diff
@@ -52,11 +68,42 @@ def initialise_database(self, file):
...
+
+            # Update versions DB mocking
+            self.initialise_database("init_version_{}".format(version))
+
+            if int(version) == 9:
+                sqlThread().create_function()
+
+            # Test versions
+            upgrade_db = UpgradeDB()
+            upgrade_db._upgrade_one_level_sql_statement(int(version))
```
Comment (PeterSurda) [inline src/tests/test_sqlthread.py:97]: ```  # pylint: disable=protected-access```

File: src/class_sqlThread.py [hunk 10]
```diff
@@ -29,22 +18,120 @@
...
+            self._upgrade_one_level_sql_statement(l)
+
+    def increment_settings_version(self, level):
+        """
+            Update version with one level
+        """
+        item = '''update settings set value=? WHERE key='version';'''
+        parameters = (level + 1,)
+        self.cur.execute(item, parameters)
+
```
Comment (PeterSurda) [inline src/class_sqlThread.py:121]: this should be more transparent. For example, you can use a `@property` decorator for properties `sql_schema_version` and `configparser_schema_version` and use these instead calling the methods directly. And setter can enforce a `+1` irrespective of what you pass as an argument.

**Comment by PeterSurda** [comment]
- rebase
- make the versions into properties
- rename `versioning` into something else, inside the methods you can use `version`, as a decorator perhaps `sql_schema_upgrade`
- delete obsolete stuff
- I haven't thoroughly reviewed tests but for starters they are probably ok
- rename the version 1 init/test to version 3. I checked the original source and it may break things if you're upgrading from version 1. If you just rename it to 3, then it would only break if we stopped the upgrade at version 2, which isn't a practical problem. You just need to make sure version 2 and 3 tests still succeeds.

**Review by PeterSurda — CHANGES_REQUESTED** [review]
Please make the requested changes.

**Comment by PeterSurda** [comment]
It is certainly an improvement but still work left to do.

**Comment by cis-muzahid** [comment]
Closing because having changes in #1794
